crates/bevy_ecs/src/schedule_v3/config.rs

```diff
@@ -0,0 +1,618 @@
+use bevy_utils::prelude::default;
+use bevy_utils::HashSet;
+
+use crate::{
+    schedule_v3::{
+        condition::{BoxedCondition, Condition},
+        graph::{Ambiguity, DependencyEdgeKind, GraphInfo},
+        set::{BoxedSystemSet, IntoSystemSet, SystemSet},
+    },
+    system::{BoxedSystem, IntoSystem, System},
+};
+
+/// A [`SystemSet`] with scheduling metadata.
+pub struct SystemSetConfig {
+    pub(super) set: BoxedSystemSet,
+    pub(super) graph_info: GraphInfo,
+    pub(super) conditions: Vec<BoxedCondition>,
+}
+
+/// A [`System`] with scheduling metadata.
+pub struct SystemConfig {
+    pub(super) system: BoxedSystem,
+    pub(super) graph_info: GraphInfo,
+    pub(super) conditions: Vec<BoxedCondition>,
+}
+
+pub(super) fn new_set_unchecked(set: BoxedSystemSet) -> SystemSetConfig {
+    SystemSetConfig {
+        set,
+        graph_info: GraphInfo {
+            sets: HashSet::new(),
+            dependencies: HashSet::new(),
+            ambiguous_with: default(),
+        },
+        conditions: Vec::new(),
+    }
+}
+
+fn new_set(set: BoxedSystemSet) -> SystemSetConfig {
+    assert!(!set.is_system_type());
```

Comment:
This assertion needs a comment: it's not immediately clear why we need to check it.